ops/main.py

```diff
@@ -156,6 +156,73 @@ def _get_event_args(charm, bound_event):
     return [], {}
 
 
+class _JujuEvent:
+    """Encapsulate how to figure out what event juju wants us to run
+
+    Also knows how to run 'legacy' hooks when juju called us via dispatch.
+
+    Args:
+        charm_dir: the toplevel directory of the charm
+
+    Attributes:
+        name: the name of the event to run
+        is_dispatch: did juju call dispatch (as opposed to a hook symlink)
+    """
+    def __init__(self, charm_dir: Path):
+        self._charm_dir = charm_dir
+        self._exec_path = Path(sys.argv[0])
+        if self._exec_path.is_symlink() and self._exec_path.name != 'dispatch':
+            self._init_legacy()
+        else:
+            self._init_dispatch()
+
+    def run_any_legacy_hook(self):
+        if not self.is_dispatch:
+            # we *are* the legacy hook
+            return
+
+        dispatch_path = self._charm_dir / self._dispatch_path
+        if not dispatch_path.exists():
+            # nothing to do
+            return
+
+        if dispatch_path.resolve() == self._exec_path.resolve():
+            # we'd just be calling ourselves
+            return
+
+        argv = sys.argv.copy()
+        argv[0] = str(dispatch_path)
+        try:
+            subprocess.run(argv, check=True)
+        except subprocess.CalledProcessError as e:
+            logger.warning("hook %s exited with status %d", dispatch_path, e.returncode)
+            sys.exit(e.returncode)
+
+    def _set_name_from_path(self, path: Path):
+        name = path.name.replace('-', '_')
+        if path.parent.name == 'actions':
+            name = '{}_action'.format(name)
+        self.name = name
+
+    def _init_legacy(self):
+        self.is_dispatch = False
+        self._set_name_from_path(self._exec_path)
+
+    def _init_dispatch(self):
+        if 'JUJU_DISPATCH_PATH' not in os.environ:
+            logger.critical("charm called via dispatch but no JUJU_DISPATCH_PATH set")
+            sys.exit(1)
+        if 'OPERATOR_DISPATCH' in os.environ:
+            # we called ourselves; bail
+            # this can't happen but it doesn't hurt to double-check
```

Comment:
"this can't happen" doesn't seem correct. Because if dispatch is a script, *and* 'install' as a hook exists, then I think we end up here, don't we?
dispatch_path.resolve() won't == hook_path.resolve()

Reply:
removed the comment (and added logging)